### PR TITLE
8340232: Optimize DataInputStream::readUTF

### PR DESCRIPTION
**Pull request by wenshao**
Similar to ObjectInputStream, use JLA.countPositives and JLA.inflateBytesToChars to speed up readUTF

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8340232](https://bugs.openjdk.org/browse/JDK-8340232): Optimize DataInputStream::readUTF (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20903/head:pull/20903` \
`$ git checkout pull/20903`

Update a local copy of the PR: \
`$ git checkout pull/20903` \
`$ git pull https://git.openjdk.org/jdk.git pull/20903/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20903`

View PR using the GUI difftool: \
`$ git pr show -t 20903`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20903.diff">https://git.openjdk.org/jdk/pull/20903.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20903#issuecomment-2354434062)

### CODE REVIEW

File: src/java.base/share/classes/java/io/DataInputStream.java [hunk 1]
```diff
@@ -592,12 +597,15 @@ public static final String readUTF(DataInput in) throws IOException {
         int chararr_count=0;
 
         in.readFully(bytearr, 0, utflen);
+        int ascii = JLA.countPositives(bytearr, 0, utflen);
+        if (ascii == utflen) {
+            return new String(bytearr, 0, utflen, StandardCharsets.ISO_8859_1);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/io/DataInputStream.java:602]: Maybe use `sun.nio.cs.ISO_8859_1.INSTANCE` instead:
```suggestion
            return new String(bytearr, 0, utflen, ISO_8859_1.INSTANCE);
```

Comment (liach) [reply src/java.base/share/classes/java/io/DataInputStream.java:602]: We should use `StandardCharsets` as in most other standard usages. This method is not on bootstrap path, so readability is more important.

**Comment by liach** [comment]
Another comment: we currently eagerly allocate a char array, but now we can choose to only allocate the char array if we have non-ascii.

**Comment by wenshao** [comment]
In the readUTFAscii scenario, the performance is significantly improved, but in the readUTFSmall scenario, the performance is regressed.

Here are the performance numbers on the MacBook M1 Pro:

## Script

```sh
# baseline
git checkout fbe2629303bcee5855673b7e37d8c49f19dc9849
make test TEST="micro:java.io.DataInputStreamTest.readUTF"

# current
git checkout 5c4bd6db977ca8729e078d26cb7ca284ab42203c
make test TEST="micro:java.io.DataInputStreamTest.readUTF"

````

## Performance Numbers

```diff
-# baseline
-Benchmark                              Mode  Cnt  Score   Error  Units
-DataInputStreamTest.readUTFAsciiLarge  avgt   20  1.401 ? 0.005  us/op
-DataInputStreamTest.readUTFAsciiMixed  avgt   20  1.337 ? 0.092  us/op
-DataInputStreamTest.readUTFAsciiSmall  avgt   20  0.872 ? 0.004  us/op
-DataInputStreamTest.readUTFLarge       avgt   20  2.178 ? 0.050  us/op
-DataInputStreamTest.readUTFMixed       avgt   20  1.700 ? 0.109  us/op
-DataInputStreamTest.readUTFSmall       avgt   20  0.879 ? 0.012  us/op

+# current
+Benchmark                              Mode  Cnt  Score   Error  Units
+DataInputStreamTest.readUTFAsciiLarge  avgt   20  0.833 ? 0.004  us/op
+DataInputStreamTest.readUTFAsciiMixed  avgt   20  0.804 ? 0.001  us/op
+DataInputStreamTest.readUTFAsciiSmall  avgt   20  0.828 ? 0.005  us/op
+DataInputStreamTest.readUTFLarge       avgt   20  2.053 ? 0.063  us/op
+DataInputStreamTest.readUTFMixed       avgt   20  1.675 ? 0.056  us/op
+DataInputStreamTest.readUTFSmall       avgt   20  1.834 ? 0.026  us/op
```

|   | baseline  | current | delta |
| --- | --- | --- | --- |
| DataInputStreamTest.readUTFAsciiLarge | 1.401 | 0.833 | 68.19% |
| DataInputStreamTest.readUTFAsciiMixed | 1.337 | 0.804 | 66.29% |
| DataInputStreamTest.readUTFAsciiSmall | 0.872 | 0.828 | 5.31% |
| DataInputStreamTest.readUTFLarge | 2.178 | 2.053 | 6.09% |
| DataInputStreamTest.readUTFMixed | 1.700 | 1.675 | 1.49% |
| DataInputStreamTest.readUTFSmall | 0.879 | 1.834 | -52.07% |

File: src/java.base/share/classes/java/io/DataInputStream.java [hunk 2]
```diff
@@ -573,31 +578,51 @@ public final String readUTF() throws IOException {
...
             }
-            chararr = dis.chararr;
-            bytearr = dis.bytearr;
-        } else {
+        }
+        boolean allocate = false;
+        if (bytearr == null) {
             bytearr = new byte[utflen];
-            chararr = new char[utflen];
+            allocate = true;
```
[the remark this answers is not in the document]
Comment (wenshao) [reply src/java.base/share/classes/java/io/DataInputStream.java:590]: When ascii != utflen, bytearr will be reused, and the name of `trusted` is not clear here.

[the remark this answers is not in the document]
Comment (liach) [reply src/java.base/share/classes/java/io/DataInputStream.java:590]: I mean to add a `trusted = false;` when bytearr is reused; trusted will be clear there, and in the future it's less likely for programmers to accidentally leak the trusted array.

**Review by liach — APPROVED** [review]
Thanks for the logical cleanup. I will find another io reviewer too.

/reviewers 2 reviewer

**Review by bplb — APPROVED** [review]

**Comment by wenshao** [comment]
\integrate

**Pull request closed without merging**